ext/Draw/CDraw3DLine.cpp

```diff
@@ -0,0 +1,75 @@
+#include "../ext/Draw/CDraw3DLine.h"
+#include "../ext/Draw/Draw3DLineShaders.h"
+
+using namespace irr;
+using namespace video;
+using namespace scene;
+using namespace ext;
+using namespace draw;
+
+CDraw3DLine* CDraw3DLine::create(IVideoDriver* _driver)
+{
+    return new CDraw3DLine(_driver);
+}
+
+CDraw3DLine::CDraw3DLine(IVideoDriver* _driver)
+:   m_driver(_driver),
+    m_desc(_driver->createGPUMeshDataFormatDesc()),
+    m_meshBuffer(new IGPUMeshBuffer())
+{
+    auto callBack = new Draw3DLineCallBack();
+    m_material.MaterialType = (E_MATERIAL_TYPE)
+        m_driver->getGPUProgrammingServices()->addHighLevelShaderMaterial(
+        Draw3DLineVertexShader,
+        nullptr,nullptr,nullptr,
+        Draw3DLineFragmentShader,
+        3,EMT_SOLID,
+        callBack,
+        0);
+    callBack->drop();
+}
+
+void CDraw3DLine::draw(
+    float fromX, float fromY, float fromZ,
+    float toX, float toY, float toZ,
+    std::uint32_t r, std::uint32_t g, std::uint32_t b, std::uint32_t a)
+{
+    S3DLineVertex vertices[2] = {
+        {{ fromX, fromY, fromZ }, { r, g, b, a }},
+        {{ toX, toY, toZ }, { r, g, b, a }}
+    };
+
+    auto upStreamBuff = m_driver->getDefaultUpStreamingBuffer();
+    m_lineData[0] = vertices;
```

Comment:
`m_lineData` needs not be a class member

Reply:
for multi-threading capability ... right.

Reply:
then m_offsets should not be a member either

Reply:
not for MT capability, it just simply doesn't need to be kept around outside of the function.

for MT its best to create one instance per-thread and use these independently...

Reply:
commiting this